File: screens/progress_screen.py
import flet as ft
import flet_charts as fch

from services.db_service import (
    get_streak_days,
    get_minutes_this_week,
    get_average_fluency,
    get_minutes_per_day_last_7,
    get_recent_fluency_scores,
)
from theme import colors

ACHIEVEMENTS = [
    ("local_fire_department", "Racha activa", "Segui practicando cada dia"),
    ("mic", "Primera conversacion", "Completaste tu reto mas grande"),
]


def build_stat_card(value: str, label: str) -> ft.Container:
    return ft.Container(
        expand=True,
        bgcolor=colors.BG_CARD,
        border_radius=12,
        padding=14,
        content=ft.Column(
            horizontal_alignment=ft.CrossAxisAlignment.CENTER,
            spacing=2,
            controls=[
                ft.Text(value, size=20, weight=ft.FontWeight.W_500, color=colors.TEXT_PRIMARY),
                ft.Text(label, size=10, color=colors.TEXT_SECONDARY, text_align=ft.TextAlign.CENTER),
            ],
        ),
    )


def build_bar_chart() -> fch.BarChart:
    data = get_minutes_per_day_last_7()
    max_val = max((minutes for _, minutes in data), default=1) or 1
    groups = []
    for i, (day_label, minutes) in enumerate(data):
        is_peak = minutes == max_val and minutes > 0
        groups.append(
            fch.BarChartGroup(
                x=i,
                rods=[
                    fch.BarChartRod(
                        from_y=0,
                        to_y=minutes,
                        width=16,
                        color=colors.ORANGE_STREAK if is_peak else colors.BLUE_ACCENT,
                        border_radius=4,
                    )
                ],
            )
        )

    return fch.BarChart(
        groups=groups,
        height=140,
        bottom_axis=fch.ChartAxis(
            labels=[
                fch.ChartAxisLabel(value=i, label=ft.Text(day_label, size=11, color=colors.TEXT_SECONDARY))
                for i, (day_label, _) in enumerate(data)
            ],
            label_size=20,
        ),
        left_axis=fch.ChartAxis(show_labels=False),
        border=ft.Border.all(0, ft.Colors.TRANSPARENT),
        interactive=False,
    )


def build_line_chart() -> fch.LineChart:
    scores = get_recent_fluency_scores(6)
    points = [fch.LineChartDataPoint(i, score) for i, score in enumerate(scores)]
    return fch.LineChart(
        data_series=[
            fch.LineChartData(
                points=points,
                stroke_width=2,
                color=colors.BLUE_ACCENT,
                curved=True,
                rounded_stroke_cap=True,
            )
        ],
        height=140,
        left_axis=fch.ChartAxis(show_labels=False),
        bottom_axis=fch.ChartAxis(show_labels=False),
        border=ft.Border.all(0, ft.Colors.TRANSPARENT),
        interactive=False,
        min_y=max(0, min(scores) - 5),
        max_y=max(scores) + 5,
    )


def build_achievement_card(icon_name: str, title: str, subtitle: str) -> ft.Container:
    return ft.Container(
        bgcolor=colors.BG_CARD,
        border_radius=12,
        padding=12,
        content=ft.Row(
            spacing=12,
            vertical_alignment=ft.CrossAxisAlignment.CENTER,
            controls=[
                ft.Container(
                    width=36,
                    height=36,
                    border_radius=18,
                    bgcolor=colors.BG_CARD_ICON,
                    alignment=ft.Alignment.CENTER,
                    content=ft.Icon(getattr(ft.Icons, icon_name.upper()), size=18, color=colors.BLUE_ACCENT),
                ),
                ft.Column(
                    spacing=1,
                    controls=[
                        ft.Text(title, size=13, weight=ft.FontWeight.W_500, color=colors.TEXT_PRIMARY),
                        ft.Text(subtitle, size=11, color=colors.TEXT_SECONDARY),
                    ],
                ),
            ],
        ),
    )


def build_progress_screen(page: ft.Page) -> ft.Container:
    return ft.Container(
        expand=True,
        bgcolor=colors.BG_PAGE,
        padding=ft.Padding.only(left=20, right=20, top=20),
        content=ft.Column(
            expand=True,
            spacing=16,
            scroll=ft.ScrollMode.AUTO,
            controls=[
                ft.Column(
                    spacing=2,
                    controls=[
                        ft.Text("TU EVOLUCION", size=12, weight=ft.FontWeight.W_500, color=colors.BLUE_ACCENT),
                        ft.Text("Progreso", size=24, weight=ft.FontWeight.W_500, color=colors.TEXT_PRIMARY),
                    ],
                ),
                ft.Row(
                    spacing=12,
                    controls=[
                        build_stat_card(str(get_streak_days()), "DIAS SEGUIDOS"),
                        build_stat_card(str(get_minutes_this_week()), "MIN. ESTA SEMANA"),
                        build_stat_card(f"{get_average_fluency()}%", "FLUIDEZ"),
                    ],
                ),
                ft.Container(
                    bgcolor=colors.BG_CARD,
                    border_radius=16,
                    padding=16,
                    content=ft.Column(
                        spacing=10,
                        controls=[
                            ft.Text("Minutos por dia", size=13, weight=ft.FontWeight.W_500, color=colors.TEXT_PRIMARY),
                            build_bar_chart(),
                        ],
                    ),
                ),
                ft.Container(
                    bgcolor=colors.BG_CARD,
                    border_radius=16,
                    padding=16,
                    content=ft.Column(
                        spacing=10,
                        controls=[
                            ft.Text(
                                "Puntaje de fluidez - ultimas sesiones",
                                size=13,
                                weight=ft.FontWeight.W_500,
                                color=colors.TEXT_PRIMARY,
                            ),
                            build_line_chart(),
                        ],
                    ),
                ),
                ft.Column(
                    spacing=10,
                    controls=[build_achievement_card(icon, title, subtitle) for icon, title, subtitle in ACHIEVEMENTS],
                ),
            ],
        ),
    )

File: app/src/main/java/com/example/serena/ui/screens/ProgressScreen.kt
package com.example.serena.ui.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.DarkMode
import androidx.compose.material.icons.filled.Mic
import androidx.compose.material.icons.filled.WaterDrop
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.example.serena.data.AppDatabase
import com.example.serena.data.PracticeSessionEntity
import com.example.serena.data.ProgressCalculator
import com.example.serena.ui.theme.*

private data class Achievement(val icon: androidx.compose.ui.graphics.vector.ImageVector, val title: String, val subtitle: String, val unlocked: Boolean)

@Composable
fun ProgressScreen(onOpenJournal: () -> Unit) {
    val context = LocalContext.current
    val db = remember { AppDatabase.getInstance(context) }
    var sessions by remember { mutableStateOf(listOf<PracticeSessionEntity>()) }

    LaunchedEffect(Unit) {
        sessions = db.practiceSessionDao().getAll()
    }

    val streakDays = ProgressCalculator.getStreakDays(sessions)
    val minutesThisWeek = ProgressCalculator.getMinutesThisWeek(sessions)
    val averageFluency = ProgressCalculator.getAverageFluency(sessions)
    val minutesPerDay = ProgressCalculator.getMinutesPerDayLast7(sessions)
    val fluencyScores = ProgressCalculator.getRecentFluencyScores(sessions)

    val achievements = listOf(
        Achievement(Icons.Filled.WaterDrop, "3 dias seguidos", "Constancia inicial desbloqueada", streakDays >= 3),
        Achievement(Icons.Filled.Mic, "Primera conversacion", "Completaste tu primer ejercicio de habla", sessions.isNotEmpty()),
        Achievement(Icons.Filled.DarkMode, "10 sesiones totales", "Ya llevas 10 practicas en Serena", sessions.size >= 10)
    ).filter { it.unlocked }

    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(BgPage)
            .padding(horizontal = 20.dp, vertical = 16.dp)
    ) {
        Text("TU EVOLUCION", color = BlueAccent, fontSize = 11.sp)
        Text("Progreso", color = TextPrimary, fontSize = 22.sp, fontWeight = FontWeight.Bold)
        Spacer(modifier = Modifier.height(14.dp))

        Row(horizontalArrangement = Arrangement.spacedBy(10.dp)) {
            StatCard(streakDays.toString(), "DIAS SEGUIDOS", Modifier.weight(1f))
            StatCard(minutesThisWeek.toString(), "MIN. ESTA SEMANA", Modifier.weight(1f))
            StatCard("$averageFluency%", "FLUIDEZ", Modifier.weight(1f))
        }
        Spacer(modifier = Modifier.height(14.dp))

        Box(
            modifier = Modifier
                .fillMaxWidth()
                .background(BgCard, shape = RoundedCornerShape(14.dp))
                .clickable { onOpenJournal() }
                .padding(14.dp),
            contentAlignment = Alignment.Center
        ) {
            Text("Ver diario de bloqueos", color = BlueAccent, fontSize = 13.sp, fontWeight = FontWeight.Bold)
        }
        Spacer(modifier = Modifier.height(14.dp))

        Column(
            modifier = Modifier
                .fillMaxWidth()
                .weight(1f)
                .background(BgCard, shape = RoundedCornerShape(14.dp))
                .padding(14.dp)
        ) {
            Text("Minutos por dia", color = TextPrimary, fontSize = 12.sp, fontWeight = FontWeight.Bold)
            Spacer(modifier = Modifier.height(10.dp))
            BarChart(data = minutesPerDay, modifier = Modifier.fillMaxWidth().weight(1f))
        }
        Spacer(modifier = Modifier.height(14.dp))

        Column(
            modifier = Modifier
                .fillMaxWidth()
                .weight(1f)
                .background(BgCard, shape = RoundedCornerShape(14.dp))
                .padding(14.dp)
        ) {
            Text("Puntaje de fluidez - ultimas sesiones", color = TextPrimary, fontSize = 12.sp, fontWeight = FontWeight.Bold)
            Spacer(modifier = Modifier.height(10.dp))
            LineChart(scores = fluencyScores, modifier = Modifier.fillMaxWidth().weight(1f))
        }
        Spacer(modifier = Modifier.height(14.dp))

        if (achievements.isEmpty()) {
            Box(
                modifier = Modifier
                    .fillMaxWidth()
                    .background(BgCard, shape = RoundedCornerShape(14.dp))
                    .padding(16.dp),
                contentAlignment = Alignment.Center
            ) {
                Text(
                    "Todavia no tenes logros. Segui practicando para desbloquear el primero.",
                    color = TextSecondary,
                    fontSize = 12.sp,
                    textAlign = TextAlign.Center
                )
            }
        } else {
            Column(verticalArrangement = Arrangement.spacedBy(10.dp)) {
                achievements.forEach { achievement ->
                    Row(
                        modifier = Modifier
                            .fillMaxWidth()
                            .background(BgCard, shape = RoundedCornerShape(14.dp))
                            .padding(14.dp),
                        verticalAlignment = Alignment.CenterVertically
                    ) {
                        Box(
                            modifier = Modifier
                                .size(36.dp)
                                .background(BgCardIcon, shape = CircleShape),
                            contentAlignment = Alignment.Center
                        ) {
                            Icon(achievement.icon, contentDescription = null, tint = BlueAccent, modifier = Modifier.size(18.dp))
                        }
                        Spacer(modifier = Modifier.width(12.dp))
                        Column {
                            Text(achievement.title, color = TextPrimary, fontSize = 13.sp, fontWeight = FontWeight.Bold)
                            Text(achievement.subtitle, color = TextSecondary, fontSize = 11.sp)
                        }
                    }
                }
            }
        }
    }
}

@Composable
private fun StatCard(value: String, label: String, modifier: Modifier = Modifier) {
    Column(
        modifier = modifier
            .background(BgCard, shape = RoundedCornerShape(10.dp))
            .padding(10.dp),
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        Text(value, color = TextPrimary, fontSize = 17.sp, fontWeight = FontWeight.Bold)
        Text(label, color = TextSecondary, fontSize = 8.sp, textAlign = TextAlign.Center)
    }
}

@Composable
private fun BarChart(data: List<Pair<String, Int>>, modifier: Modifier = Modifier) {
    val maxValue = (data.maxOfOrNull { it.second } ?: 1).coerceAtLeast(1).toFloat()
    Row(modifier = modifier, horizontalArrangement = Arrangement.spacedBy(6.dp)) {
        data.forEach { (label, minutes) ->
            val isPeak = minutes.toFloat() == maxValue && minutes > 0
            Column(
                modifier = Modifier.weight(1f).fillMaxHeight(),
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.Bottom
            ) {
                Box(
                    modifier = Modifier
                        .fillMaxWidth()
                        .fillMaxHeight((minutes / maxValue).coerceIn(0.02f, 1f))
                        .background(if (isPeak) OrangeStreak else BlueAccent, shape = RoundedCornerShape(3.dp))
                )
                Spacer(modifier = Modifier.height(4.dp))
                Text(label, color = TextSecondary, fontSize = 9.sp)
            }
        }
    }
}

@Composable
private fun LineChart(scores: List<Int>, modifier: Modifier = Modifier) {
    val minScore = (scores.minOrNull() ?: 0).toFloat()
    val maxScore = (scores.maxOrNull() ?: 100).toFloat()
    val range = (maxScore - minScore).coerceAtLeast(1f)

    androidx.compose.foundation.Canvas(modifier = modifier) {
        if (scores.size < 2) return@Canvas
        val stepX = size.width / (scores.size - 1)
        val points = scores.mapIndexed { index, score ->
            val x = stepX * index
            val normalized = (score - minScore) / range
            val y = size.height - (normalized * size.height * 0.85f) - (size.height * 0.05f)
            androidx.compose.ui.geometry.Offset(x, y)
        }

        val linePath = androidx.compose.ui.graphics.Path().apply {
            points.forEachIndexed { index, point ->
                if (index == 0) moveTo(point.x, point.y) else lineTo(point.x, point.y)
            }
        }

        val fillPath = androidx.compose.ui.graphics.Path().apply {
            addPath(linePath)
            lineTo(points.last().x, size.height)
            lineTo(points.first().x, size.height)
            close()
        }

        drawPath(
            path = fillPath,
            brush = androidx.compose.ui.graphics.Brush.verticalGradient(
                colors = listOf(BlueAccent.copy(alpha = 0.35f), BlueAccent.copy(alpha = 0f))
            )
        )
        drawPath(
            path = linePath,
            color = BlueAccent,
            style = androidx.compose.ui.graphics.drawscope.Stroke(width = 3f, cap = androidx.compose.ui.graphics.StrokeCap.Round)
        )
        points.dropLast(1).forEach { point ->
            drawCircle(color = BlueAccent, radius = 3f, center = point)
        }
        drawCircle(color = OrangeStreak, radius = 5f, center = points.last())
    }
}